utils/resolved_udp_conn.go

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) 2020 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"context"
+	"fmt"
+	"github.com/uber/jaeger-client-go/log"
+	"net"
+	"sync"
+	"time"
+)
+
+// ResolvedUDPConn is an implementation of udpConn that continually resolves the provided hostname and atomically swaps
+// the connection if the dial succeeds
+type ResolvedUDPConn struct {
+	hostPort    string
+	resolveFunc resolveFunc
+	dialFunc    dialFunc
+	logger      log.Logger
+	bufferBytes int
+
+	connMtx   sync.RWMutex
+	conn      *net.UDPConn
+	destAddr  *net.UDPAddr
+	closeChan chan struct{}
+	doneChan  chan struct{}
+}
+
+type resolveFunc func(network string, hostPort string) (*net.UDPAddr, error)
+type dialFunc func(network string, laddr, raddr *net.UDPAddr) (*net.UDPConn, error)
+
+// NewResolvedUDPConn returns a new udpConn that resolves hostPort every resolveTimeout, if the resolved address is
+// different than the current conn then the new address is dial and the conn is swapped
+func NewResolvedUDPConn(hostPort string, resolveTimeout time.Duration, resolveFunc resolveFunc, dialFunc dialFunc, logger log.Logger) (*ResolvedUDPConn, error) {
+	conn := &ResolvedUDPConn{
+		hostPort:    hostPort,
+		resolveFunc: resolveFunc,
+		dialFunc:    dialFunc,
+		logger:      logger,
+	}
+
+	err := conn.attemptResolveAndDial()
+	if err != nil {
```

Comment:
Returning the error here would typically make the application exit. Some people complained about race conditions between application startup and jaeger agent startup. Since this connection keeps retrying, I wonder if it's better to not return an error here and just keep retrying. After all, the same error can occur later in the resolveLoop.

Of course, it might need additional logic to deal with "connection not initialized" situation. Just a food for thought, doesn't need to be part of this PR.

Reply:
I've made the change to not return an error if first resovle+dial fails. If a write comes in and no connection is initialized an attempt is made to establish one, if that succeeds the write is tried. One caveat of this strategy is that the retry would be coupled with bufferFlushInterval . What do you think?